usr/share/rear/lib/bareos-functions.sh

```diff
@@ -0,0 +1,301 @@
+    #
+    # debian11-fd Version: 21.1.10 (04 June 2024)  Debian GNU/Linux 11 (bullseye)
+    # Daemon started 13-Jun-24 08:50. Jobs: run=0 running=0, Bareos subscription binary
+    #  Sizeof: boffset_t=8 size_t=8 debug=0 trace=0 bwlimit=0kB/s
+    #
+    # Running Jobs:
+    # bareos-dir (director) connected at: 13-Jun-24 13:41
+    # No Jobs running.
+    # ====
+    #
+    # Terminated Jobs:
+    #  JobId  Level    Files      Bytes   Status   Finished        Name 
+    # ======================================================================
+    #     18  Full     59,653    2.504 G  OK       05-Jun-24 14:00 backup-client1
+    # ====
+
+    # When bconsole can't access the Bareos Director, it exits with an error code.
+    # When the Director does not know the client, it prints an error message and no "Connecting to Client".
+    # When the Director knows about the client, but can't reach it,
+    # the "Running Jobs:" text does not appear.
+    # The "Running Jobs:" headline even shown when no jobs are running.
+    # In this case, the additional sentence "No Jobs running." is added.
+
+    local rc
+    local bconsole_client_status
+    bconsole_client_status=$(bconsole <<< "status client=$client")
+    rc=$?
+    if [ $rc -eq 0 ]; then
+        Log "${bconsole_client_status}"
+    else
+        LogPrint "${bconsole_client_status}"
+        Error "Failed to connect to Bareos Director."
+    fi
+    LogPrint "Connecting to the Bareos Director: OK"
+
+    if ! grep "Connecting to Client $client" <<< "$bconsole_client_status"; then
+        Error "Failure: The Bareos Director cannot connect to the local filedaemon ($client)."
+    fi
+
+    if ! grep "Running Jobs:" <<< "${bconsole_client_status}"; then
+        Error "Failure: The Bareos Director cannot connect to the local filedaemon ($client)."
+    fi
+
+    LogPrint "Bareos Director: can connect to the local filedaemon ($client)"
+}
+
+function get_available_restore_job_names()
+{
+    # example output of 'bcommand_json "show jobs"':
+    # {
+    #   "jsonrpc": "2.0",
+    #   "id": null,
+    #   "result": {
+    #     "jobs": {
+    #       "RestoreFiles": {
+    #         "name": "RestoreFiles",
+    #         "type": "Restore",
+    #         "fileset": "LinuxAll",
+    #         "where": "/tmp/bareos-restores",
+    #         "jobdefs": "DefaultJob"
+    #       },
+    #       "backup-bareos-fd": {
+    #         "name": "backup-bareos-fd",
+    #         "client": "bareos-fd",
+    #         "jobdefs": "DefaultJob"
+    #       }
+    #     }
+    #   }
+    # }
+    local show_jobs
+    show_jobs="$( bcommand_json "show jobs" )"
+    jq --exit-status --raw-output '.result.jobs | with_entries(select(.value.type == "Restore")) | .[].name' <<< "$show_jobs"
+}
+
+function get_last_restore_jobid()
+{
+    # example output of 'bcommand_json "list jobs client=client1-fd jobtype=R last"':
+    # {
+    #   "jsonrpc": "2.0",
+    #   "id": null,
+    #   "result": {
+    #     "jobs": [
```

Comment:
Empty result is fine. Then the result will be also empty. Because of the "last" parameter, I expected that only one result is possible, however because of your remark I though about this again, and in some cases there can be multiple results. I change the code to pick the maximum jobid instead.

Reply:
Maximum means latest? Can we be sure that that is the "right" one?

Maybe instead of making such assumptions we can simply pass the job ID of the actual restore job into these functions and use that? I'd hope that the command that starts the restore job will get the restore job ID as a result and then we can use that to be more explicit and use less guesswork.

Reply:
Yes, maximum means latest. I think, this is reliable. However, this is now only used to be of help in manual mode. In automatic mode it gets the JobId from the restore command result.